src/modules/hip/kernel/non_linear_blend.hpp

```diff
@@ -0,0 +1,286 @@
+
+    float4 rowLocComponent_f4 = (float4)rowLocComponent;
+    float4 multiplier_f4 = (float4)multiplier;
+
+    d_float8 colLocComponent_f8;
+    colLocComponent_f8.f4[0] = make_float4(idXY_i2.x, idXY_i2.x + 1, idXY_i2.x + 2, idXY_i2.x + 3);
+    colLocComponent_f8.f4[1] = colLocComponent_f8.f4[0] + (float4)4;
+    colLocComponent_f8.f4[0] -= (float4)halfDimsWH_i2.x;
+    colLocComponent_f8.f4[1] -= (float4)halfDimsWH_i2.x;
+    colLocComponent_f8.f4[0] = (colLocComponent_f8.f4[0] * colLocComponent_f8.f4[0] * multiplier_f4) + rowLocComponent_f4;
+    colLocComponent_f8.f4[1] = (colLocComponent_f8.f4[1] * colLocComponent_f8.f4[1] * multiplier_f4) + rowLocComponent_f4;
+
+    gaussianValue_f8->f4[0] = make_float4(expf(colLocComponent_f8.f4[0].x), expf(colLocComponent_f8.f4[0].y), expf(colLocComponent_f8.f4[0].z), expf(colLocComponent_f8.f4[0].w));
+    gaussianValue_f8->f4[1] = make_float4(expf(colLocComponent_f8.f4[1].x), expf(colLocComponent_f8.f4[1].y), expf(colLocComponent_f8.f4[1].z), expf(colLocComponent_f8.f4[1].w));
+}
+
+__device__ void non_linear_blend_8_hip_compute(d_float8 *src1_f8, d_float8 *src2_f8, d_float8 *dst_f8, d_float8 *gaussianValue_f8)
+{
+    dst_f8->f4[0] = (src1_f8->f4[0] - src2_f8->f4[0]) * gaussianValue_f8->f4[0] + src2_f8->f4[0];
+    dst_f8->f4[1] = (src1_f8->f4[1] - src2_f8->f4[1]) * gaussianValue_f8->f4[1] + src2_f8->f4[1];
+}
+
+__device__ void non_linear_blend_24_hip_compute(d_float24 *src1_f24, d_float24 *src2_f24, d_float24 *dst_f24, d_float8 *gaussianValue_f8)
+{
+    non_linear_blend_8_hip_compute(&(src1_f24->f8[0]), &(src2_f24->f8[0]), &(dst_f24->f8[0]), gaussianValue_f8);
+    non_linear_blend_8_hip_compute(&(src1_f24->f8[1]), &(src2_f24->f8[1]), &(dst_f24->f8[1]), gaussianValue_f8);
+    non_linear_blend_8_hip_compute(&(src1_f24->f8[2]), &(src2_f24->f8[2]), &(dst_f24->f8[2]), gaussianValue_f8);
+}
+
+template <typename T>
+__global__ void non_linear_blend_pkd_tensor(T *srcPtr1,
+                                            T *srcPtr2,
+                                            uint2 srcStridesNH,
+                                            T *dstPtr,
+                                            uint2 dstStridesNH,
+                                            float *stdDev,
+                                            RpptROIPtr roiTensorPtrSrc)
+{
+    int id_x = (hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x) * 8;
+    int id_y = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+
+    if ((id_y >= roiTensorPtrSrc[id_z].xywhROI.roiHeight) || (id_x >= roiTensorPtrSrc[id_z].xywhROI.roiWidth))
+    {
+        return;
+    }
+
+    uint srcIdx = (id_z * srcStridesNH.x) + ((id_y + roiTensorPtrSrc[id_z].xywhROI.xy.y) * srcStridesNH.y) + ((id_x + roiTensorPtrSrc[id_z].xywhROI.xy.x) * 3);
+    uint dstIdx = (id_z * dstStridesNH.x) + (id_y * dstStridesNH.y) + id_x * 3;
+
+    float multiplier = -0.5f / (stdDev[id_z] * stdDev[id_z]);
```

Comment:
Is there a way to avoid this floating point division inside the kernel?

Reply:
It's hard since stdDev is supplied by the user. In the short term, we could always do a host loop outside computing (-0.5f/(stdDev*stdDev)), followed by conversion to device mem.
But in the long term, if RPP will be expecting uniformly all params to be in GPU mem too, then this has to ideally be done inside the kernel.

Reply:
Yes. I think we need to figure a way to do this more efficiently in future version